tests/testapp/tests/test_api.py

```diff
@@ -0,0 +1,791 @@
+
+        # delete two of the certs from the chain so we can make sure they get added back
+        self.sub_subset_cert1_with_key.delete()
+        self.subset_cert1_without_key.delete()
+        self.assertEqual(Certificate.objects.count(), self.original_cert_count - 2)
+
+        # make the API call to create the SyncSession
+        response = self.client.post(reverse('syncsessions-list'), data, format='json')
+        self.assertEqual(response.status_code, 201)
+
+        # check that the cert chain was deserialized
+        self.assertEqual(Certificate.objects.count(), self.original_cert_count)
+
+        # check that the syncsession was created
+        syncsession = SyncSession.objects.get()
+        self.assertEqual(syncsession.id, data["id"])
+        self.assertEqual(syncsession.remote_certificate_id, data["client_certificate_id"])
+        self.assertEqual(syncsession.local_certificate_id, data["server_certificate_id"])
+        self.assertTrue(syncsession.active)
+
+    def test_syncsession_creation_fails_with_bad_signature(self):
+
+        data = self.get_initial_syncsession_data_for_request()
+
+        data["signature"] = self.sub_subset_cert1_with_key.sign("nonsense:id")
+
+        self.assertSyncSessionCreationFails(data)
+
+    def test_syncsession_creation_fails_with_client_cert_not_matching_cert_chain(self):
+
+        data = self.get_initial_syncsession_data_for_request()
+
+        data["certificate_chain"] = json.dumps(CertificateSerializer(self.subset_cert2_with_key.get_ancestors(include_self=True), many=True).data)
+
+        self.assertSyncSessionCreationFails(data)
+
+    def test_syncsession_creation_fails_with_expired_nonce(self):
+
+        data = self.get_initial_syncsession_data_for_request()
+
+        Nonce.objects.all().update(timestamp=timezone.datetime(2000, 1, 1, tzinfo=timezone.get_current_timezone()))
+
+        self.assertSyncSessionCreationFails(data)
+
+    def test_syncsession_creation_fails_with_nonexistent_nonce(self):
+
+        data = self.get_initial_syncsession_data_for_request()
+
+        Nonce.objects.all().delete()
+
+        self.assertSyncSessionCreationFails(data)
+
+    def test_syncsession_creation_fails_with_nonexistent_server_certificate(self):
+
+        data = self.get_initial_syncsession_data_for_request()
+
+        data["server_certificate_id"] = uuid.uuid4().hex
+
+        self.assertSyncSessionCreationFails(data, status_code=400)
+
+    def test_syncsession_can_be_deleted(self):
+
+        self.test_syncsession_can_be_created()
+
+        syncsession = SyncSession.objects.get()
+        self.assertEqual(syncsession.active, True)
+
+        response = self.client.delete(reverse('syncsessions-detail', kwargs={"pk": syncsession.id}), format='json')
+        self.assertEqual(response.status_code, 204)
+
+        # check that the syncsession was "deleted"
+        self.assertEqual(SyncSession.objects.get().active, False)
```

Comment:
If 0 records exist, it blows up. If > 1 records exist, it blows up. Otherwise, it returns the single record that does exist.

Will change though, yep, if it makes it clearer.